File: stack-queue/py/Stack.py

class Node:
    def __init__(self,val = None ,next = None):
        self.val = val
        self.next = next


class Stack:
    def __init__(self,val= None):
        self.head = self.tail = Node(val) if val is not None else None
        self.len = 1 if val is not None else 0


    def push(self,val):

        node = Node(val)
        
        if self.head is None:
           self.head = self.tail = node
        else:
           self.tail.next = node
           self.tail = node

        self.len += 1


    def pop(self):
     
        if self.head is None:
            print("Stack is empty")
            return
        
        if self.len == 1:
            self.head = self.tail = None
            self.len = 0
            return
        
        previous_node = self.find_node(self.len - 1)
        self.tail = previous_node
        self.tail.next = None
        self.len -= 1 


    def find_node(self,index):
        
        if index < 1 or index > self.len:
          print("index is out of range")
          return
        
        data = self.head
        num = 1

        while data:
           
           if num == index:
              break
           
           num += 1
           data = data.next

        return data
     
     


my_stack = Stack()
my_stack.push(10)
my_stack.push(20)
my_stack.pop()
my_stack.pop()
my_stack.pop()
print("head",my_stack.head)
print("tail",my_stack.tail)
print("len",my_stack.len)

File: stack-queue/js/Stack.js
class Node {
  constructor(value, next = null) {
    this.value = value;
    this.next = next;
  }
}

class Stack {
  constructor(value) {
    this.head = this.tail = value ? new Node(value) : null;
    this.length = value ? 1 : 0;
  }

  push(value) {
    const node = new Node(value);

    if (!this.head) {
      this.head = this.tail = node;
    } else {
      this.tail.next = node;
      this.tail = node;
    }
    this.length++;
  }

  pop() {
    if (!this.head) {
      console.log("Stack is empty");
      return;
    }

    if (this.length - 1 === 0) {
      this.head = this.tail = null;
      this.length = 0;
      return;
    }

    let previousNode = this.findNode(this.length - 1);
    this.tail = previousNode;
    this.tail.next = null;
    this.length--;
  }

  findNode(index) {
    if (index < 1 || index > this.length) {
      console.log("index out of bounds");
      return;
    }

    let data = this.head;
    let count = 1;

    while (data) {
      if (count === index) {
        break;
      }
      count++;
      data = data.next;
    }

    return data;
  }
}

const myStack = new Stack();
myStack.push(10);
myStack.push(20);
myStack.pop();
myStack.pop();

console.log(myStack);
